### ISSUE TITLE
Race condition in update

### ISSUE THREAD

**Issue by leedm777** [balderdashy/sails-mysql#253]
Reopening #189, since it doesn't appear to be resolved.

There is a race condition in `update` (in `adapter.js`) that I've run into a few times. The current update implementation works by:

1. [Query with the given criteria to get a list of ids](https://github.com/balderdashy/sails-mysql/blob/331f6bbb1bcd7aa534d08f6844ea1c67ebe90e7e/lib/adapter.js#L968)
2. [Update with the given criteria](https://github.com/balderdashy/sails-mysql/blob/331f6bbb1bcd7aa534d08f6844ea1c67ebe90e7e/lib/adapter.js#L1003)
3. [Query with the list of ids from step 1 to get the updated objects](https://github.com/balderdashy/sails-mysql/blob/331f6bbb1bcd7aa534d08f6844ea1c67ebe90e7e/lib/adapter.js#L1025)

The problem is that if other updates happen between steps 1 and 2, the data returned in step 3 could be inaccurate (either omitting/including records that were/were not affected).

Since this is a MySQL specific driver, there's a [pretty clever way](http://stackoverflow.com/a/1751282/115478) of capturing the ids for updated records.

**Comment by mikermcneil** [comment]
@leedm777 thanks for reopening, and thanks for the [link](http://stackoverflow.com/questions/1388025/how-to-get-id-of-the-last-updated-row-in-mysql/1751282#1751282).  This is definitely something I'd like to see a PR for-- we're not currently using MySQL ourselves, but I suspect there's a similar situation in postgresql that we could tackle ourselves as a template cc @devinivy @particlebanana

**Comment by mikermcneil** [comment]
@leedm777 added to the Sails core ROADMAP.md so we have this centralized, and in case anyone reading that has time to help (https://github.com/balderdashy/sails/commit/9c587b7749d3253b88bb101b5fc6c9643314ae02)

**Comment by devinivy** [comment]
I like that.  @tjwebb are knex or your new adapters aware of this?

**Comment by particlebanana** [comment]
@mikermcneil the postgres adapter wouldn't have this issue because it can return the values from the update. This would be useful and would cut a lot out of the update function.

It would look something like this: https://github.com/balderdashy/sails-postgresql/blob/master/lib/adapter.js#L981-L989 with `SET @uids := null;` appended to the beginning of the generated query and `AND ( SELECT @uids := CONCAT_WS(',', pk, @uids) );
SELECT @uids;` appended to the end.

**Comment by sailsbot** [comment]
Thanks for posting, @leedm777.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Reopening because this is a pending bug.

**Comment by mikermcneil** [comment]
@devinivy thank you

**Comment by mikermcneil** [comment]
@devinivy this should be good now-- the "bug" label is ignored as of Nov 16

**Comment by mikermcneil** [comment]
@leedm777 Thanks again for reopening this issue.

@shamasis is this one of the issues you guys recently addressed/encountered in your fork?

**Comment by mikermcneil** [comment]
By the way @leedm777, I'm working on cleaning up the roadmap, so if you wouldn't mind adding any other specific suggestions you might have as far as an approach for a patch, I'd appreciate it.  I took a look at  your example query in the link-- if we could implement this as a single query, this wouldn't depend on transactions, and it could be implemented as an improvement to this module.  I'm going to leave it as a pending proposal either way for the time being-- if it turns out this _cannot_ be implemented in a single query, would one of you guys open a PR to Waterline adding a backlog item to [ROADMAP.md](https://github.com/balderdashy/waterline/blob/master/ROADMAP.md)?

**Comment by shamasis** [comment]
This is generic transaction specific issue and IMHO not a race condition. :-)

**Comment by particlebanana** [comment]
Yeah because these all happen on the same connection it's probably possible to wrap this in a transaction even without adding support for transactions at a higher level.

**Comment by leedm777** [comment]
I don't think transactions alone would help; at least not with the default isolation level.

From the [MySQL docs](http://dev.mysql.com/doc/refman/5.7/en/innodb-consistent-read.html):

> The snapshot of the database state applies to SELECT statements within a transaction, not necessarily to DML statements. If you insert or modify some rows and then commit that transaction, a DELETE or UPDATE statement issued from another concurrent REPEATABLE READ transaction could affect those just-committed rows, even though the session could not query them. If a transaction does update or delete rows committed by a different transaction, those changes do become visible to the current transaction. For example, you might encounter a situation like the following:

```sql
SELECT COUNT(c1) FROM t1 WHERE c1 = 'xyz'; -- Returns 0: no rows match.
DELETE FROM t1 WHERE c1 = 'xyz'; -- Deletes several rows recently committed by other transaction.

SELECT COUNT(c2) FROM t1 WHERE c2 = 'abc'; -- Returns 0: no rows match.
UPDATE t1 SET c2 = 'cba' WHERE c2 = 'abc'; -- Affects 10 rows: another txn just committed 10 rows with 'abc' values.
SELECT COUNT(c2) FROM t1 WHERE c2 = 'cba'; -- Returns 10: this txn can now see the rows it just updated.
```

**Comment by sailsbot** [comment]
Thanks for posting, @leedm777.  I'm a repo bot-- nice to meet you!

The issue queue in this repo is for verified bugs with documented features.  Unfortunately, we can't leave some other types of issues marked as "open".  This includes bug reports about undocumented features or unofficial plugins, as well as feature requests, questions, and commentary about the core framework. Please review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) for details.

If you're here looking for help and can't find it, visit [StackOverflow](http://stackoverflow.com), our [Google Group](https://groups.google.com/forum/#!forum/sailsjs) or our [chat room](https://gitter.im/balderdashy/sails?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).  But please do not let this disrupt future conversation in this issue! I am only marking it as "closed" to keep organized.

Thanks so much for your help!

> If I am mistaken, and this issue is about a _critical bug with a documented feature_, please accept my sincerest apologies-- I am but a humble bot working to make GitHub a better place.  If you open a new issue, a member of the core team will take a look ASAP.

**Comment by magicdawn** [comment]
Status ?

**Comment by magicdawn** [comment]
@leedm777 

Will a transaction + `SELECT ... FOR UPDATE` work?
since other select will block

**Comment by leedm777** [comment]
@magicdawn That would catch some cases, but not all. The select would only lock the records that match the where clause at the time of the select. If other records are inserted/updated, they could be included in the update but have been omitted from the original select.

**Comment by magicdawn** [comment]
@leedm777 Gotcha

**Comment by magicdawn** [comment]
To get the updated ids. I tried the stackoverflow answer: 

```sql
UPDATE TAB_NAME
SET foo = 1
WHERE (
  ... -- condition
) AND (
  SELECT @update_ids := CONCAT_WS(',', @update_ids, 'id')
)
```

It's Ok, but the `LIMIT 10` will not work, if I just want to update only 10 rows
So the assignment must be placed in the `SET ...` statement

I tried to assign a variable and do not influence any innocent column
```sql
UPDATE TAB_NAME
SET foo = 1, id = ELT(1, id, CONCAT_WS(',', @update_ids, 'id'))
WHERE (
  ... -- condition
)
```

but seems mysql got some optimization on `ELT(1, id, assign-expression)`, the `assign-expression` did not execute

And finally, I add a new column for this table, called `extra`
```sql
UPDATE TAB_NAME
SET foo = 1, extra = CONCAT_WS(',', @update_ids, 'id')
WHERE (
  ... -- condition
)
```

It workes well, but need extra column. Posting this.
